### PR TITLE
Dev 2033 add updated leverage value

### PR DESCRIPTION
**Pull request by cryptokage1996**

### CODE REVIEW

**Comment by cosmic-vagabond** [comment]
@cryptokage1996 please fix conflicts and add test cases for both long and short positions thanks

File: x/perpetual/client/cli/query_mtp_test.go [hunk 1]
```diff
@@ -64,6 +64,7 @@ func networkWithMTPObjects(t *testing.T, n int) (*network.Network, []*types.MtpA
 			},
 			TradingAssetPrice: sdk.ZeroDec(),
 			Pnl:               sdk.ZeroDec(),
+			UpdatedLeverage:   sdk.ZeroDec(),
```
Comment (cosmic-vagabond) [inline x/perpetual/client/cli/query_mtp_test.go:67]: @cryptokage1996 can you share a test result where the new `updated leverage` field return a positive value for an existing position thanks

File: x/perpetual/keeper/updated_leverage.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/elys-network/elys/x/perpetual/types"
+)
+
+func (k Keeper) UpdatedLeverage(ctx sdk.Context, mtp types.MTP) (sdk.Dec, error) {
+	entry, found := k.assetProfileKeeper.GetEntry(ctx, ptypes.BaseCurrency)
+	if !found {
+		return sdk.ZeroDec(), errorsmod.Wrapf(assetprofiletypes.ErrAssetProfileNotFound, "asset %s not found", ptypes.BaseCurrency)
+	}
+	baseCurrency := entry.Denom
+	collateral_in_usdc := math.LegacyDec(mtp.Collateral)
```
Comment (cosmic-vagabond) [inline x/perpetual/keeper/updated_leverage.go:18]: @cryptokage1996 long position’s collateral can be in terms of both base currency or trading asset, we have to make sure it is in base currency

**Review by cosmic-vagabond — CHANGES_REQUESTED** [review]

File: x/perpetual/keeper/updated_leverage.go [hunk 3]
```diff
@@ -16,6 +16,10 @@ func (k Keeper) UpdatedLeverage(ctx sdk.Context, mtp types.MTP) (sdk.Dec, error)
 	}
 	baseCurrency := entry.Denom
 	collateral_in_usdc := math.LegacyDec(mtp.Collateral)
+	if mtp.CollateralAsset != baseCurrency {
+		price := k.amm.EstimatePrice(ctx, mtp.CustodyAsset, baseCurrency)
```
Comment (cosmic-vagabond) [inline x/perpetual/keeper/updated_leverage.go:20]: @cryptokage1996 price is queried twice line 20 and 25 can we remove this redundancy? thanks

**Review by cosmic-vagabond — CHANGES_REQUESTED** [review]

File: x/perpetual/keeper/updated_leverage.go (hunk 3 above)
Comment (cosmic-vagabond) [reply x/perpetual/keeper/updated_leverage.go:20]: @cryptokage1996 can you also fix the conflicts in this PR thanks

File: x/perpetual/keeper/updated_leverage.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/elys-network/elys/x/perpetual/types"
+)
+
+func (k Keeper) UpdatedLeverage(ctx sdk.Context, mtp types.MTP) (sdk.Dec, error) {
+	entry, found := k.assetProfileKeeper.GetEntry(ctx, ptypes.BaseCurrency)
+	if !found {
+		return sdk.ZeroDec(), errorsmod.Wrapf(assetprofiletypes.ErrAssetProfileNotFound, "asset %s not found", ptypes.BaseCurrency)
+	}
+	baseCurrency := entry.Denom
+	collateral_in_usdc := math.LegacyDec(mtp.Collateral)
```
Comment (avkr003) [inline x/perpetual/keeper/updated_leverage.go:18]: incorrect code, you are type casting Into to Dec, Use ToLegacyDec

File: x/perpetual/keeper/updated_leverage.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	if mtp.CollateralAsset != baseCurrency {
+		price := k.amm.EstimatePrice(ctx, mtp.CollateralAsset, baseCurrency)
+		collateral_in_usdc = math.LegacyDec(mtp.Liabilities).Mul(price)
+	}
+	liablites := math.LegacyDec(mtp.Liabilities)
```
Comment (avkr003) [inline x/perpetual/keeper/updated_leverage.go:23]: same as above, please see if you have done this anywhere else and correct it

**Review by avkr003 — CHANGES_REQUESTED** [review]

**Comment by cosmic-vagabond** [comment]
@cryptokage1996 there are multiple unit tests that are failing

**Comment by Wgil** [comment]
lets not merge this before https://github.com/elys-network/elys/pull/867

**Review by avkr003 — CHANGES_REQUESTED** [review]

File: x/perpetual/keeper/updated_leverage.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+	}
+	liablites := mtp.Liabilities.ToLegacyDec()
+	if mtp.LiabilitiesAsset != baseCurrency {
+		price := k.amm.EstimatePrice(ctx, mtp.LiabilitiesAsset, baseCurrency)
+		liablites = mtp.Liabilities.ToLegacyDec().Mul(price)
+	}
+	if collateral_in_usdc.IsZero() {
+		return sdk.ZeroDec(),  nil
+	}
+	updated_leverage := liablites.Quo(collateral_in_usdc).Add(sdk.OneDec())
```
Comment (avkr003) [inline x/perpetual/keeper/updated_leverage.go:29]: Change from 1 + liabilities/collateral to custody / (custody - liabilites)

**Review by avkr003 — APPROVED** [review]

File: proto/elys/perpetual/query.proto [hunk 7]
```diff
@@ -127,11 +127,15 @@ message MtpAndPrice {
     (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
     (gogoproto.nullable) = false
   ];
-  string liquidation_price = 4 [
+  string updated_leverage = 4 [
```
[the remark this answers is not in the document]
Comment (avkr003) [reply proto/elys/perpetual/query.proto:130]: based on new formula, it should be `effective_leverage` as we won't be using collateral

**Review by avkr003 — CHANGES_REQUESTED** [review]

File: x/perpetual/types/expected_keepers.go [hunk 8]
```diff
@@ -143,6 +143,8 @@ type AssetProfileKeeper interface {
 	GetEntry(ctx sdk.Context, baseDenom string) (val atypes.Entry, found bool)
 	// GetEntryByDenom returns a entry from its denom value
 	GetEntryByDenom(ctx sdk.Context, denom string) (val atypes.Entry, found bool)
+	// GetUsdcDenom returns value of usdc denom
+	GetUsdcDenom(ctx sdk.Context) (usdcDenom string, found bool)
```
Comment (avkr003) [inline x/perpetual/types/expected_keepers.go:147]: Can remove this if not being used

File: x/perpetual/keeper/updated_leverage.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+func (k Keeper) UpdatedLeverage(ctx sdk.Context, mtp types.MTP) (sdk.Dec, error) {
+
+	entry, found := k.assetProfileKeeper.GetEntry(ctx, ptypes.BaseCurrency)
+	if !found {
+		return sdk.ZeroDec(), errorsmod.Wrapf(assetprofiletypes.ErrAssetProfileNotFound, "asset %s not found", ptypes.BaseCurrency)
+	}
+	baseCurrency := entry.Denom
+
+	priceBaseCurrency := k.oracleKeeper.GetAssetPriceFromDenom(ctx, baseCurrency)
```
Comment (avkr003) [inline x/perpetual/keeper/updated_leverage.go:21]: No need. Base currency is usdc so price is 1

File: x/perpetual/keeper/updated_leverage.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+	if priceBaseCurrency.IsZero() {
```
Comment (avkr003) [inline x/perpetual/keeper/updated_leverage.go:22]: It's never 0

File: x/perpetual/keeper/updated_leverage.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+	custodyInUsdc := mtp.Custody.ToLegacyDec()
+	if mtp.CustodyAsset != baseCurrency {
+		priceCustodyAsset := k.oracleKeeper.GetAssetPriceFromDenom(ctx, mtp.CustodyAsset)
+		if priceBaseCurrency.IsZero() {
+			return sdk.ZeroDec(), fmt.Errorf("token price not set: %s", mtp.CustodyAsset)
+		}
+		custodyInUsdc = mtp.Custody.ToLegacyDec().Mul(priceCustodyAsset.Quo(priceBaseCurrency))
+	}
+	denominator := custodyInUsdc.Sub(mtp.Liabilities.ToLegacyDec())
```
Comment (avkr003) [inline x/perpetual/keeper/updated_leverage.go:34]: Incorrect. For short, liabilities will be trading in asset, can't subtract different units

File: x/perpetual/keeper/updated_leverage.go [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+	if mtp.LiabilitiesAsset != baseCurrency {
+		priceCustodyAsset := k.oracleKeeper.GetAssetPriceFromDenom(ctx, mtp.LiabilitiesAsset)
+		if priceBaseCurrency.IsZero() {
+			return sdk.ZeroDec(), fmt.Errorf("token price not set: %s", mtp.LiabilitiesAsset)
+		}
+		custodyInUsdc = mtp.Custody.ToLegacyDec().Mul(priceCustodyAsset.Quo(priceBaseCurrency))
+	}
```
Comment (avkr003) [inline x/perpetual/keeper/updated_leverage.go:41]: Too complicated. Check in other places how tradingAssetPrice is fetched.

**Review by avkr003 — CHANGES_REQUESTED** [review]

**Review by avkr003 — APPROVED** [review]

**Pull request merged**